python/pyiceberg/avro/resolver.py

```diff
@@ -135,7 +136,10 @@ def struct(self, struct: StructType, field_results: List[Writer]) -> Writer:
         return StructWriter(tuple(field_results))
 
     def field(self, field: NestedField, field_result: Writer) -> Writer:
-        return field_result
+        if field.required:
+            return field_result
+        else:
+            return OptionWriter(field_result)
```

Comment:
Without this change the avro file structure was the same as for a required field, which can't be parsed as an optional field.